File: Day280/binaryTreesWithFactors.go
package main

import (
	"fmt"
	"sort"
)
func numFactoresBinaryTrees(arr []int) int {
	mod := 1000000007
	sort.Ints(arr)
	a := make([]int, len(arr))
	for i,_ := range a {
		a[i] =1
	}
	m := make(map[int]int)
	for i,v := range arr {
		m[v] = i
	}
	for i:=0; i<len(arr); i++ {
		for j:=0; j<i; j++ {
			if arr[i] % arr[j] == 0 {
				temp := arr[i] /arr[j];
				if _, flag := m[temp]; flag {
					a[i] = (a[i]+a[j]*a[m[temp]])%mod
				}
			}
		}
	}
	answer := 0
	for _,v := range a {
		answer = (answer + v)%mod;
	}
	return answer
}
func main() {
	arr1 := []int{2,4}
	arr2 := []int{2,4,5,10}
	arr3:= []int{18,3,6,2}
	fmt.Println(numFactoresBinaryTrees(arr1));
	fmt.Println(numFactoresBinaryTrees(arr2));
	fmt.Println(numFactoresBinaryTrees(arr3));

}

File: Day280/binaryTreesWithFactors.java
package Day280;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class binaryTreesWithFactors {
    public static int numFactoresBinaryTrees(int arr[]) {
        int mod = 1000000007;
        Arrays.sort(arr);
        long[] a = new long[arr.length];
        Arrays.fill(a, 1);
        Map<Integer, Integer> m = new HashMap<>();
        for (int i = 0; i < arr.length; i++) {
            m.put(arr[i], i);
        }
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < i; j++) {
                if (arr[i] % arr[j] == 0) {
                    int temp = arr[i] / arr[j];
                    if (m.containsKey(temp)) {
                        a[i] = (int) ((a[i] + (long) a[j] * a[m.get(temp)]) % mod);
                    }
                }
            }
        }
        int answer = 0;
        for (long v : a) {
            answer = (int) ((answer + v) % mod);
        }
        return answer;
    }
    public static void main(String[] args) {
        int arr1[] = {2,4};
        int arr2[] = {2,4,5,10};
        int arr3[] = {18,3,6,2};
        System.out.println(numFactoresBinaryTrees(arr1));
        System.out.println(numFactoresBinaryTrees(arr2));
        System.out.println(numFactoresBinaryTrees(arr3));
    
    }
}